[NJT] Actually inline NT torch function during dynamo

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121445
* #122797

In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to @mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

soulitzer committed:
Actually inline NT torch function during dynamo

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer commented:
Re: @jbschlosser's comment from https://github.com/pytorch/pytorch/pull/121375#issuecomment-1989115915

> This certainly explains the symptom we were seeing during AOTAutograd. Perhaps we should clean up the terminology a bit; we have:
> - "Traceable wrapper subclasses" - used throughout Dynamo / AOTAutograd; wrapper subclasses that implement torch_dispatch and are decomposable into inner tensors. These are defined by implementing the __tensor_flatten__() / __tensor_unflatten__() APIs appropriately.
> - "Traceable tensor subclasses" - separate set of __torch_function__() subclasses that Dynamo may or may not trace into. These are defined by membership within the traceable_tensor_subclasses set.

I would definitely agree that it might be worth clarifying the naming here. cc @mlazos @bdhirsh

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -468,9 +468,12 @@ def create_aot_dispatcher_function(
+            fake_mode, \
+            python_dispatcher_mode, \
+            PhiloxStateTracker(), \
+            torch._C._DisableTorchFunctionNonInfraSubclass():
```
It seems to me that this should be toggleable? Because if I directly call aot autograd (e.g., from non-strict export), I would *not* expect the torch function subclass to have already be inlined and therefore I really should not disable it.

Overall, I find doing the disabling here inside AOTAutograd a bit smelly, but I'm not sure what to do about it. Potentially, it could be that *Dynamo* should be responsible for disabling non-infra torch function subclasses before calling into the backend, as a way of indicating that it has already inlined through them. So you'd move this over to output_graph.py or something.

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer commented on torch/fx/proxy.py:
```diff
@@ -370,6 +370,9 @@ class Proxy:
     """
+    # Indicates to our torch_function dispatching infra that
+    # this is an "infra" object
+    _mode_key = torch._C._TorchFunctionModeKey.PROXY
```
After the short-term fix, I don't actually have a test case that verifies this is doing something...

soulitzer commented:
TODO:
- add additional testing for other sdpa paths to the PR (today only flash is tested to not graph break), this might reveal other things we need to fix
- investigate other CI brokenness

ezyang commented on test/test_nestedtensor.py:
```diff
@@ -3912,13 +3911,14 @@ def test_sdpa_compile(self, device, dtype):
         self.assertEqual(attn_d1, attn_nts[0].unsqueeze(0), atol=output_ref_atol, rtol=output_ref_rtol)
         self.assertEqual(attn_d2, attn_nts[1].unsqueeze(0), atol=output_ref_atol, rtol=output_ref_rtol)
 
+    @xfailIfTorchDynamo
```
You know what's up with this one?

soulitzer replied:
I'm not sure what is causing this yet, but I found a different issue that seems quite important to fix: 

`torch._dynamo.exc.Unsupported: Accessing overridden method/attribute values on a tensor subclass with a __torch_function__ override is not supported`

It makes everywhere we access `.values()` now graph break:
e.g., `output_dense = F.scaled_dot_product_attention(query.values(), key.values(), value.values())`

cc @mlazos

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1585,9 +1591,20 @@ def _clone_input(value):
+        torch.sym_sqrt,
+    ]:
+        proxy.node.meta["example_value"] = example_value
+        return ConstantVariable.create(example_value, **options)
```
uhhh you sure???

soulitzer replied:
- Oh right, we want to constant fold this and never reach this function (as per the comment above the function!)
- I'm going to try to use normal `math.sqrt` instead of sym_sqrt and see what happens. Apparently there is a comment there about normal sqrt not working, but I'm not sure why yet.

ezyang reviewed (commented):
This seems fine but there's one howler in the Dynamo code?

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer commented:
For clarity: the current status of this is that to land this we still need additional support on the torch function <> compile end to avoid regressing with more graph breaks.

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "[NJT] Actually inline NT torch function during dynamo"


In order to avoid having any temporary state where the behavior of anything is regressed. This PR does all of the following at once:

(1) Disables torch function running a second time in AOTAutograd

If you have a tensor subclass that relies on dispatch into the same op without unwrapping and calling torch._C.DisableTorchFunctionSubclass() the torch function-ness will survive into AOTAutograd (when normally we may expect the torch function to be inlined away during dynamo). If this happens, we should make sure to not run the torch function logic a second time.

2.  Enables torch function to be inlined in dynamo for NT

Due to torch function running a second time AOTAutograd, NT was actually relying on this behavior instead of properly inlining through torch function at the dynamo level. 

3. Fixes graph breaks for NT torch function

Now that we are inlining through torch function for the first time in dynamo, we've uncovered some graph breaks. Thanks to mlazos, we should have support for custom attributes for torch function now. We also add support for a custom Enum type. Finally, a few of them we can get rid of by adding allow_in_graph (though we may need to double check the soundness here).


Fixes https://github.com/pytorch/pytorch/issues/120654, https://github.com/pytorch/pytorch/issues/120124

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]
